include/rc_runtime.h

```diff
@@ -132,12 +136,12 @@ typedef struct rc_runtime_event_t {
 }
 rc_runtime_event_t;
 
-typedef void (*rc_runtime_event_handler_t)(const rc_runtime_event_t* runtime_event);
+typedef void (*rc_runtime_event_handler_t)(const rc_runtime_event_t* runtime_event, void* ud);
 
-void rc_runtime_do_frame(rc_runtime_t* runtime, rc_runtime_event_handler_t event_handler, rc_runtime_peek_t peek, void* ud, lua_State* L);
+void rc_runtime_do_frame(rc_runtime_t* runtime, rc_runtime_event_handler_t event_handler, rc_runtime_peek_t peek, lua_State* L);
 void rc_runtime_reset(rc_runtime_t* runtime);
 
-typedef int (*rc_runtime_validate_address_t)(unsigned address);
```

Comment:
Would rc_runtime_ud be intended to be public with this code or purely internal? In the latter case, I would assume the API would (contractually) disallow usage of rc_runtime_get_userdata while using rc_runtime_validate_addresses (while being perfectly fine with the extended method). Although in such a case, why store the old userdata in the struct since that's only used again after the ex function does work (better store it as a local variable).

Reply:
`rc_runtime_ud` would be internal in this implementation. The legacy `rc_runtime_validate_addresses` wouldn't be expected to call a function using `rc_runtime_get_userdata`. Newer implementations using `rc_runtime_set_userdata` would call `rc_runtime_validate_addresses_ex` with the new handler that gets passed the `runtime->userdata` value they provided.

Reply:
I would suppose it could be possible to ensure userdata is not pointing to our internal userdata, like so
```c
typedef struct rc_compat_validate_ud {
  rc_runtime_event_handler_t event_handler;
  rc_runtime_validate_address_t validate_handler;
  void* userdata;
}
rc_compat_validate_ud;

static void rc_runtime_event_handler_translate(rc_runtime_event_t* runtime_event) {
  rc_compat_validate_ud* ud = (rc_compat_validate_ud*)runtime_event->runtime->userdata;
  runtime_event->runtime->userdata = ud->userdata;
  ud->event_handler(runtime_event);
  runtime_event->runtime->userdata = ud;
}

static void rc_runtime_validate_address_translate(unsigned address, rc_runtime_t* runtime) {
  rc_compat_validate_ud* ud = (rc_compat_validate_ud*)runtime->userdata;
  ud->validate_handler(address);
}

void rc_runtime_validate_addresses(rc_runtime_t* self, rc_runtime_event_handler_t event_handler, 
    rc_runtime_validate_address_t validate_handler) {
  rc_compat_validate_ud compat_validate_ud;
  compat_validate_ud.event_handler = event_handler;
  compat_validate_ud.validate_handler = validate_handler;
  compat_validate_ud.userdata = self->userdata;
  self->userdata = &compat_validate_ud;
  rc_runtime_validate_addresses_ex(self, rc_runtime_event_handler_translate, rc_runtime_validate_address_translate);
  self->userdata = compat_validate_ud.userdata;
}
```

However, instead switching in the old userdata, maybe it would be smarter to null it out, as to ensure a user using rc_runtime_get_userdata with the runtime_event's runtime just immediately crashes rather than letting odd code expecting the old userdata being present to "work" (either case, ensures that the user doesn't get a pointer to somewhere on the stack).

Reply:
I don't think we need a event_handler translation layer. We can add another field to the runtime_event_t without affecting existing implementations. Otherwise, that's pretty much exactly what I proposed earlier.

Reply:
Also, perhaps it could be fair to allow a mix of userdata attached event handling with no userdata attached validate handling. A validate handler might not require any kind of userdata, for emulators which only emulate a single console and have a static memory map, and therefore do not need any userdata is the validate handler (as no state is modified at all), while an event handler very much requires some state modified and therefore may want userdata.

Reply:
I think it's fine to pass user data to both functions. I'm just saying that we can't change the existing function _signatures_. We can change the `rc_runtime_event_t` structure without affecting any signatures. There's no way to pass user data to the validate handler without changing the validate handler signature.

The proposal is to turn the existing validation function into a wrapper for the new validation function that provides an internal method and data structure to redirect to a function with the old validate handler signature.